### ISSUE TITLE
Adding Oblique Trees to the Cythonized Tree Module

### ISSUE THREAD

**Issue by adam2392** [adam2392/scikit-learn#2]
This is a draft Issue to file to the actual repo. cc: @rflperry @ChesterHuynh @jovo

#### Describe the workflow you want to enable
Oblique trees originally proposed by, Breiman 2001 [1], would enable a more flexible class of decision trees. These are formed by linear combinations (possibly sparse) of the input data, which has been shown to perform better on a wide range of simulated and benchmark datasets in [2]. Oblique trees would enable sparse projection oblique random forests (SPORF). Besides empirical evidence, the sparse constraint here borrows from the Johnson Lindenstrauss theorem.

#### Describe your proposed solution
The original implementation was a pure C++ implementation that utilizes many of the lower-level features of C++ (https://sporf.neurodata.io). Moreover, more structured trees (i.e. manifold trees) that assume some manifold structure in the input data, can be built on top of the oblique tree codebase (by inheriting class structures from the Python/Cython code in scikit-learn). Some preliminary results on manifold oblique random forests are in [3], which help provide more empirical evidence that a base oblique forest class would be useful in scikit-learn.

The implementation of SPORF was ported into a scikit-learn fork (https://github.com/adam2392/scikit-learn), which was started by Parth Vora, and then finished by Chester Huynh, Adam Li and members of Joshua Vogelstein's lab. The SPORF Cython implementation currently builds cleanly on top of the existing cython code for the regular Decision Trees and thus does not affect the runtime, or performance on Random Forests at all. 

The `.pxd` files can also be included in each scikit-learn release, so that other sklearn-compatible forest modules can build upon our implementation of the Oblique tree.

#### Describe alternatives you've considered, if relevant
As with all proposals to scikit-learn, an alternative could be to just keep things in the `rerf` package. However, this would limit its extensibility and usability in the community. The idea of oblique forests has been proposed and been around since 2001, so it has in some sense stood the test of time as being a potentially useful idea in machine learning to make decision trees more flexible.

#### Relevant Benchmarks
To ease scikit-learn developer concerns on incorporating such a large piece of code change, we can provide the following benchmarks and tests on Cythonized SPORF:

1. benchmark tests showing the runtime performance on a sparse parity problem (see [2]) relevant to Random Forests 
2. benchmark tests on cc_18 vs Random Forests

#### Additional context

[1]: https://link.springer.com/article/10.1023/A:1010933404324
[2]: https://arxiv.org/abs/1506.03410

**Comment by jovo** [comment]
great start.

decouple from our work to the extent possible, and highlight Breiman's original comment about how linear combination stuff worked better.  

then, cite ROFLMAO, CCF, SPORF, and other recent oblique tree stuff.  be explicit in the citations.

do not mention the possibility of keeping it in our repo. 
i do not think we need to mention who did what.

include the benchmark figures when they are complete.

**Comment by jovo** [comment]
1. "Moreover in that paper, he did note that linear combinations performed better then random independent features." 

i wanted the quote where he said L-RC was better than F-RC. 

2. "The implementation of SPORF was ported into a scikit-learn fork (https://github.com/adam2392/scikit-learn). The SPORF Cython". 

I don't think i'd mention SPORF here at all.  just say we implemented sparse oblique forests.

3. the 2nd figure is not explained, what is delta-kappa, what does >0 mean, what about <0.  also, do we have improved results yet?

**Comment by adam2392** [comment]
> 1. "Moreover in that paper, he did note that linear combinations performed better then random independent features."
> 
> i wanted the quote where he said L-RC was better than F-RC.

Hmm I don't think the original RF paper explicitly mentions a linear-RC. Do you mean this one?

"On the larger data sets, F=8 gives better results. Forest-RC does exceptionally well on the synthetic data sets. Overall, it compares more favorably to Adaboost than Forest-RI." Where Forest-RI is our traditional random forests, and Forest-RC is a forest with linear combinations.

> I don't think i'd mention SPORF here at all. just say we implemented sparse oblique forests.

+1

> 1. the 2nd figure is not explained, what is delta-kappa, what does >0 mean, what about <0.  also, do we have improved results yet?

Working on those rn. We've managed to fix the speed and performance issue, so that cy-obliqueforests run on all cc_18 examples when they were previously unable to do so due to speed issues.

Need to run side-by-side rerf comparisons to really understand why they are different and on what datasets. Rn the code semantically looks exactly the same to us.

**Comment by jovo** [comment]
awesome, yes, that quote, excited for new results. thanks

**Comment by jovo** [comment]
edit of text

**Describe the workflow you want to enable**

We are proposing to add a cythonized module that allows for building oblique trees.

Oblique trees originally proposed by, Breiman 2001 [1], would enable a more flexible class of decision trees. Breiman called a version of these Forest-RC, as compared to what most people call random forest, which he called Forest-RI.  He noted: 

"Forest-RC does exceptionally well on the synthetic data sets. Overall, it compares more favorably to Adaboost than Forest-RI."

In other words, in Breiman's paper introduction random forest over 20 years ago, with over 75,000 citations, he presented 2 algorithms, and argued that the one which does sparse linear combinations of features worked better than the naive one which most people have used since.  More recently, several authors have corroborated these original results [2-5]. Oblique trees in general open up a wide class of flexibility for trees to take advantage of structure in data. Besides empirical evidence, the sparse constraint here borrows from the Johnson Lindenstrauss theorem.

**Describe your proposed solution**

The implementation of sparse oblique random forest  (SPORF) was ported into a scikit-learn fork (https://github.com/adam2392/scikit-learn). The SPORF Cython implementation currently builds cleanly on top of the existing cython code for the regular Decision Trees and thus does not affect the runtime, or performance on Random Forests at all. Moreover, it improves upon RF in certain cc_18 benchmarks (see figure below).

The `.pxd` files can also be included in each scikit-learn release, so that other sklearn-compatible forest modules can build upon our implementation of the Oblique tree.

**Relevant Benchmarks**

To ease scikit-learn developer concerns on incorporating such a large piece of code change, we can provide the following benchmarks and tests on Cythonized RF (cysporf). These benchmarks were ran on the OpenML CC-18 dataset. Max features were set to `n_features`, so runtime was very high for both RF and oblique forests in e.g. Fashion-MNIST.


`please compute statistical test illustrating that sporf works better than rf`
